python/symbolica_community/tensors.pyi

```diff
@@ -0,0 +1,307 @@
+        body. For example the function `f(x,y)=x^2+y` should be provided as
+        `{(f, "f", (x, y)): x**2 + y}`. All free parameters should be provided in the `params` list.
+        """
+
+    def scalar(self)->Expression:
+        r"""
+        Unwrap the tensor into a scalar, if possible.
+        """
+
+class TensorEvaluator:
+    r"""
+    An optimized evaluator for tensors.
+    """
+    def evaluate(self,inputs:Sequence[Sequence[float]])-> List[Tensor]:
+        """Evaluate the expression for multiple inputs and return the result."""
+
+    def evaluate_complex(self, inputs: Sequence[Sequence[complex]]) -> List[Tensor]:
+        """Evaluate the expression for multiple complex inputs and return the result."""
+
+    def compile(
+        self,
+        function_name: str,
+        filename: str,
+        library_name: str,
+        inline_asm: bool = True,
+        optimization_level: int = 3,
+        compiler_path: Optional[str] = None,
+    ) -> CompiledTensorEvaluator:
+        """Compile the evaluator to a shared library using C++ and optionally inline assembly and load it."""
+
+class TensorIndices(Sequence):
+    r"""
+    A structure that can be used to represent the "shape" of a tensor, along with a list of abstract indices.
+    This has an optional name, and accompanying symbolica expressions that are considered as additional non-indexed arguments.
+    The structure is essentially a list of `Slots` that are used to define the structure of the tensor.
+    """
+
+    def __new__(cls,name:Expression,*additional_args:Slot|Expression) -> TensorIndices:
+        r"""
+        Create a new tensor structure with the given name and slots. Any additional expressions are also passed along.
+        """
+
+    def __repr__(self) -> str:
+        ...
+
+    def __str__(self) -> str:
+        ...
+
+    def to_expression(self) -> Expression:
+        r"""
+        Return the symbolica expression of the tensor indices.
+        """
+
+    def __len__(self) -> int:
+        r"""
+        Return the size of the tensor.
+        """
+
+    @overload
+    def __getitem__(self, idx: int) -> List[int]:
+        r"""
+        Get the indices of the slot at the given flat index.
+        """
+
+    @overload
+    def __getitem__(self, s: slice) -> Sequence[List[int]]:
+        r"""
+        Get the list of expanded indices of the slot at the given slice of flat index.
+        """
+
+    @overload
+    def __getitem__(self, key: List[int]) -> int:
+        r"""
+        Get the flattened index of the given indices.
+        """
+
+class TensorNetwork:
+    r"""
+    A tensor network.
+
+    This class is a wrapper around the `TensorNetwork` class from the `spenso` crate.
+    Such a network is a graph representing the arithmetic operations between tensors.
+    In the most basic case, edges represent the contraction of indices.
+    """
+    def __new__(cls,expr:Expression) -> TensorNetwork:
```

Comment:
Could you add an example section here as this is the main entry point?

```
Examples
--------
a = TensorNetwork(....)
```

Reply:
Yes good point!